Runtime backend autodetection (#523)
dalek-cryptography/curve25519-dalek: README.md

| `alloc`            |    ✓     | Enables Edwards and Ristretto multiscalar multiplication, batch scalar inversion, and batch Ristretto double-and-compress. Also enables `zeroize`. |
| `zeroize`          |    ✓     | Enables [`Zeroize`][zeroize-trait] for all scalar and curve point types. |
| `precomputed-tables` |    ✓     | Includes precomputed basepoint multiplication tables. This speeds up `EdwardsPoint::mul_base` and `RistrettoPoint::mul_base` by ~4x, at the cost of ~30KB added to the code size. |

> | `simd_avx2`        |    ✓     | Allows the AVX2 SIMD backend to be used, if available. |
> | `simd_avx512`      |    ✓     | Allows the AVX512 SIMD backend to be used, if available. |
> | `simd`             |    ✓     | Allows every SIMD backend to be used, if available. |

Reviewer: ```suggestion
| `forbid_avx2`        |    ✓     | Disallows the AVX2 SIMD backend to be used, if available. |
| `forbid_avx512`    |    ✓     | Disallows the AVX512 SIMD backend to be used, if available. |
| `forbid_simd`        |    ✓     | Disallows every SIMD backend to be used, if available. |
```

Curve arithmetic is implemented and used by selecting one of the following backends:

> | Backend            | Implementation                                                | Target backends             |
> | :---               | :---                                                          | :---                        |
> | `[default]`        | Automatic runtime backend selection (either serial or SIMD)   | `u32` <br/> `u64` <br/> `avx2` <br/> `avx512` |
> | `fiat`             | Formally verified field arithmetic from [fiat-crypto]         | `fiat_u32` <br/> `fiat_u64` |

Reviewer: We should still document / refer to the backends in a single list.

```suggestion
| Backend            | Selection | Implementation                                                | Target backends             |
| :---               | :---                                                          | :---                        |
| `[default]`        | Automatic | Automatic runtime backend selection (either serial or SIMD)   | `u32` <br/> `u64` <br/> `avx2` <br/> `avx512` |
| `simd`             | Automatic | [Parallel][parallel_doc], using Advanced Vector Extensions | `avx2` <br/> `avx512ifma`   |
| `fiat`             | Manual | Formally verified field arithmetic from [fiat-crypto]         | `fiat_u32` <br/> `fiat_u64` |
```

> To choose a backend other than the `[default]` backend, set the

Reviewer: ```suggestion
To manually select the manually selectable `fiat` backend, set the
```

environment variable:
```sh

> RUSTFLAGS='--cfg curve25519_dalek_backend="BACKEND"'

Reviewer: ```suggestion
RUSTFLAGS='--cfg curve25519_dalek_backend="fiat"'
```

To choose a backend other than the `[default]` serial backend, set the
To choose a backend other than the `[default]` backend, set the
environment variable:
```sh
RUSTFLAGS='--cfg curve25519_dalek_backend="BACKEND"'
```
where `BACKEND` is `simd` or `fiat`. Equivalently, you can write to
where `BACKEND` is `fiat`. Equivalently, you can write to
`~/.cargo/config`:
```toml
[build]
rustflags = ['--cfg=curve25519_dalek_backend="BACKEND"']
```
More info [here](https://doc.rust-lang.org/cargo/reference/config.html#buildrustflags).

The `simd` backend requires extra configuration. See [the SIMD
section](#simd-target-backends).

> 

Reviewer: Please note that the SIMD backend may be automatically selected where these instructions are supported but the automatic SIMD backend selection can be disallowed via feature flags which will apply to whole use of curve25519-dalek. It is not advisable to use these disallow feature flags from individual library crates as this will take the control away from binary crate.

## SIMD target backends

Target backend selection within `simd` must be done manually by setting the
`RUSTFLAGS` environment variable to one of the below options:
The SIMD target backend selection is done automatically at runtime depending
on the available CPU features, provided the appropriate feature flag is enabled.

| CPU feature | `RUSTFLAGS`                     | Requires nightly? |
| :---        | :---                            | :---              |
| avx2        | `-C target_feature=+avx2`       | no                |

> | avx512ifma  | `-C target_feature=+avx512ifma` | yes               |

Reviewer 2: I noticed all of this documentation is removed, but the newly introduced crate features aren't equivalent and perhaps this should be documented.

Namely these flags *bypass autodetection* and force the use of SIMD features.

Reviewer: yeah I can take care of documenting that - just need to re-organise a bit to make room for the vendored dep first
